File: python/src/multi_agent_orchestrator/agents/amazon_bedrock_agent.py
"""
Amazon Bedrock Agent Integration Module

This module provides a robust implementation for interacting with Amazon Bedrock agents,
offering a flexible and extensible way to process conversational interactions using
AWS Bedrock's agent runtime capabilities.
"""

from typing import Any, Optional
from dataclasses import dataclass
import os
import boto3
from botocore.exceptions import BotoCoreError, ClientError
from multi_agent_orchestrator.agents import Agent, AgentOptions, AgentStreamResponse
from multi_agent_orchestrator.types import ConversationMessage, ParticipantRole
from multi_agent_orchestrator.utils import Logger
from multi_agent_orchestrator.shared import user_agent


@dataclass
class AmazonBedrockAgentOptions(AgentOptions):
    """
    Configuration options for Amazon Bedrock Agent initialization.

    Provides flexible configuration for Bedrock agent runtime:
    - agent_id: Unique identifier for the Bedrock agent
    - agent_alias_id: Specific alias for the agent
    - client: Optional custom boto3 client (allows dependency injection)
    - streaming: Flag to enable streaming response mode (on final response)
    - enableTrace: Flag to enable detailed event tracing
    """
    region: Optional[str] = None
    agent_id: str = None
    agent_alias_id: str = None
    client: Any | None = None
    streaming: bool | None = False
    enableTrace: bool | None = False


class AmazonBedrockAgent(Agent):
    """
    Specialized agent for interacting with Amazon Bedrock's intelligent agent runtime.

    This class extends the base Agent class to provide:
    - Direct integration with AWS Bedrock agent runtime
    - Configurable response handling (streaming/non-streaming)
    - Comprehensive error management
    - Flexible session and conversation state management
    """

    def __init__(self, options: AmazonBedrockAgentOptions):
        """
        Initialize the Bedrock agent with comprehensive configuration.

        Handles client creation, either using a provided client or
        automatically creating one based on AWS configuration.

        :param options: Detailed configuration for agent initialization
        """
        super().__init__(options)

        # Store core agent identifiers
        self.agent_id = options.agent_id
        self.agent_alias_id = options.agent_alias_id

        # Set up Bedrock runtime client
        if options.client:
            # Use provided client (useful for testing or custom configurations)
            self.client = options.client
        else:
            # Create default client using AWS region from options or environment
            self.client = boto3.client('bedrock-agent-runtime',
                                       region_name=options.region or os.environ.get('AWS_REGION'))

        user_agent.register_feature_to_client(self.client, feature="bedrock-agent")


        # Configure response handling modes
        self.streaming = options.streaming
        self.enableTrace = options.enableTrace

    def is_streaming_enabled(self) -> bool:
        """
        Check if streaming mode is active for response processing.

        :return: Boolean indicating streaming status
        """
        return self.streaming is True

    async def process_request(
        self,
        input_text: str,
        user_id: str,
        session_id: str,
        chat_history: list[ConversationMessage],
        additional_params: dict[str, str] | None = None
    ) -> ConversationMessage:
        """
        Process a user request through the Bedrock agent runtime.

        Handles the entire interaction lifecycle:
        - Manages session state
        - Invokes agent with configured parameters
        - Processes streaming or non-streaming responses
        - Handles potential errors

        :param input_text: User's input message
        :param user_id: Identifier for the user
        :param session_id: Unique conversation session identifier
        :param chat_history: Previous conversation messages
        :param additional_params: Optional supplementary parameters
        :return: Agent's response as a conversation message
        """
        # Initialize session state, defaulting to empty if not provided
        session_state = {}
        if (additional_params and 'sessionState' in additional_params):
            session_state = additional_params['sessionState']

        try:
            # Configure streaming behavior
            streamingConfigurations = {
                'streamFinalResponse': self.streaming
            }

            # Invoke Bedrock agent with comprehensive configuration
            response = self.client.invoke_agent(
                agentId=self.agent_id,
                agentAliasId=self.agent_alias_id,
                sessionId=session_id,
                inputText=input_text,
                enableTrace=self.enableTrace,
                sessionState=session_state,
                streamingConfigurations=streamingConfigurations if self.streaming else {}
            )

            completion = ""

            if self.streaming:
                async def generate_chunks():
                    nonlocal completion
                    for event in response['completion']:
                        if 'chunk' in event:
                            chunk = event['chunk']
                            decoded_response = chunk['bytes'].decode('utf-8')
                            self.callbacks.on_llm_new_token(decoded_response)
                            completion += decoded_response
                            yield AgentStreamResponse(text=decoded_response)
                        elif 'trace' in event and self.enableTrace:
                            Logger.info(f"Received event: {event}")
                    yield AgentStreamResponse(
                        final_message=ConversationMessage(
                            role=ParticipantRole.ASSISTANT.value,
                            content=[{'text':completion}]))
                return generate_chunks()
            else:
                for event in response['completion']:
                    if 'chunk' in event:
                        chunk = event['chunk']
                        decoded_response = chunk['bytes'].decode('utf-8')
                        self.callbacks.on_llm_new_token(decoded_response)
                        completion += decoded_response
                    elif 'trace' in event and self.enableTrace:
                        Logger.info(f"Received event: {event}")

                return ConversationMessage(
                    role=ParticipantRole.ASSISTANT.value,
                    content=[{"text": completion}]
                )
        except (BotoCoreError, ClientError) as error:
            # Comprehensive error logging and propagation
            Logger.error(f"Error processing request: {str(error)}")
            raise error

File: typescript/src/agents/amazonBedrockAgent.ts
import { BedrockAgentRuntimeClient, InvokeAgentCommand, InvokeAgentCommandOutput } from "@aws-sdk/client-bedrock-agent-runtime";
import { ConversationMessage, ParticipantRole } from "../types";
import { Agent, AgentOptions } from "./agent";
import { Logger } from "../utils/logger";

/**
 * Options for configuring an Amazon Bedrock agent.
 * Extends base AgentOptions with specific parameters required for Amazon Bedrock.
 */
export interface AmazonBedrockAgentOptions extends AgentOptions {
  region?: string;
  agentId: string;        // The ID of the Amazon Bedrock agent.
  agentAliasId: string;   // The alias ID of the Amazon Bedrock agent.
  client?: BedrockAgentRuntimeClient;  // Client for interacting with the Bedrock agent runtime.
  enableTrace?: boolean;  // Flag to enable tracing of Agent
  streaming?: boolean;    // Flag to enable streaming of responses.
}


/**
 * Represents an Amazon Bedrock agent that interacts with a runtime client.
 * Extends base Agent class and implements specific methods for Amazon Bedrock.
 */
export class AmazonBedrockAgent extends Agent {
  private agentId: string;                    // The ID of the Amazon Bedrock agent.
  private agentAliasId: string;               // The alias ID of the Amazon Bedrock agent.
  private client: BedrockAgentRuntimeClient;  // Client for interacting with the Bedrock agent runtime.
  private enableTrace: boolean;// Flag to enable tracing of Agent
  private streaming: boolean;    // Flag to enable streaming of responses.

  /**
   * Constructs an instance of AmazonBedrockAgent with the specified options.
   * Initializes the agent ID, agent alias ID, and creates a new Bedrock agent runtime client.
   * @param options - Options to configure the Amazon Bedrock agent.
   */
  constructor(options: AmazonBedrockAgentOptions) {
    super(options);
    this.agentId = options.agentId;
    this.agentAliasId = options.agentAliasId;
    this.client = options.client ? options.client : options.region
    ? new BedrockAgentRuntimeClient({ region: options.region })
    : new BedrockAgentRuntimeClient();
    this.enableTrace = options.enableTrace || false;
    this.streaming = options.streaming || false;
  }

  private async *handleStreamingResponse(response: InvokeAgentCommandOutput): AsyncIterable<string> {
    for await (const chunkEvent of response.completion) {
      if (chunkEvent.chunk) {
        const chunk = chunkEvent.chunk;
        const decodedResponse = new TextDecoder("utf-8").decode(chunk.bytes);
        yield decodedResponse;
      } else if (chunkEvent.trace){
        if (this.enableTrace){
          Logger.logger.info("Trace:", JSON.stringify(chunkEvent.trace));
        }
      }
    }
  }

  /**
   * Processes a user request by sending it to the Amazon Bedrock agent for processing.
   * @param inputText - The user input as a string.
   * @param userId - The ID of the user sending the request.
   * @param sessionId - The ID of the session associated with the conversation.
   * @param chatHistory - An array of Message objects representing the conversation history.
   * @param additionalParams - Optional additional parameters as key-value pairs.
   * @returns A Promise that resolves to a Message object containing the agent's response.
   */
  /* eslint-disable @typescript-eslint/no-unused-vars */
  async processRequest(
    inputText: string,
    userId: string,
    sessionId: string,
    chatHistory: ConversationMessage[],
    additionalParams?: Record<any, any>
  ): Promise<ConversationMessage | AsyncIterable<any>> {
    // Construct the command to invoke the Amazon Bedrock agent with user input
    const command = new InvokeAgentCommand({
      agentId: this.agentId,
      agentAliasId: this.agentAliasId,
      sessionId: sessionId,
      inputText: inputText,
      sessionState: additionalParams ? additionalParams.sessionState?  additionalParams.sessionState : undefined : undefined,
      enableTrace: this.enableTrace,
      streamingConfigurations: {
        streamFinalResponse: this.streaming,
      }
    });

    try {
      let completion = "";
      const response = await this.client.send(command);

      // Process the response from the Amazon Bedrock agent
      if (response.completion === undefined) {
        throw new Error("Completion is undefined");
      }

      if (this.streaming){
        return this.handleStreamingResponse(response);
      } else {
        // Aggregate chunks of response data
        for await (const chunkEvent of response.completion) {
          if (chunkEvent.chunk) {
            const chunk = chunkEvent.chunk;
            const decodedResponse = new TextDecoder("utf-8").decode(chunk.bytes);
            completion += decodedResponse;
          } else if (chunkEvent.trace) {
            if (this.enableTrace){
              Logger.logger.info("Trace:", JSON.stringify(chunkEvent.trace));
            }
          }
        }
      }

      // Return the completed response as a Message object
      return {
        role: ParticipantRole.ASSISTANT,
        content: [{ text: completion }],
      };
    } catch (err) {
      // Handle errors encountered while invoking the Amazon Bedrock agent
      Logger.logger.error(err);
      throw err;
    }
  }
}

